### ISSUE TITLE
tuning hyperparameters of a modified model

### ISSUE THREAD

**Issue by gorkemguzeler** [ultralytics/ultralytics#7564]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I have a modified model including different blocks in nn.module. Therefore, when i try to tune using tuner.py: I do not want to use the _ultralytics' yolo train commands_ which does not include my modifications. 2 possible approaches comes to my mind:
1) Instead of running in subprocess to avoid dataloader hang, I will train normally (such as using a DetectionTrainer, shared what i have in my mind below). Would this cause any issues?

```
                #### original ####
                # cmd = ['yolo', 'train', *(f'{k}={v}' for k, v in train_args.items())]
                # assert subprocess.run(cmd, check=True).returncode == 0, 'training failed'

                #### modified ####
                trainer = DetectionTrainer(overrides=train_args)
                assert subprocess.run(trainer.train(), check=True).returncode == 0, 'training failed'
```

2) I should use torch methods to build my own tuner but I do not know if that's doable in a short notice. 

Thanks!

Best regards,

**Comment by glenn-jocher** [comment]
@gorkemguzeler hi there,

Great to hear you're working on customizing your model with different blocks! For tuning hyperparameters of a modified model, you might consider the following:

1. **Custom Training Loop**: You can indeed replace the subprocess call with your own training loop using `DetectionTrainer`. This should work as long as your `DetectionTrainer` class correctly handles the training process. Just ensure that your data loading, training, and validation steps are properly integrated.

2. **Custom Tuner**: Building your own tuner using PyTorch methods is certainly an option. It might require more effort, but it gives you full control over the tuning process. You'll need to handle hyperparameter sampling, model training, and performance evaluation.

Remember to keep an eye on data loading and GPU utilization to avoid potential issues related to multiprocessing and distributed training.

Best of luck with your tuning, and feel free to reach out if you have further questions!

Happy coding! 😊

P.S. For more detailed guidance on training and other modes, check out our documentation at https://docs.ultralytics.com.

**Comment by gorkemguzeler** [comment]
@glenn-jocher hi, thanks a lot for your help! 

for anyone going through a similar case; i think it is best to use YOLO pipeline since it provides a lot of easiness (data loading/preprocessing, loss functions etc.). Then going over the results of each training to extract and compare the results. 

Best,

**Comment by glenn-jocher** [comment]
@gorkemguzeler You're welcome!

Absolutely, leveraging the YOLO pipeline can streamline many aspects of your workflow, from data handling to utilizing established loss functions. Iteratively training and reviewing results is a solid approach to refining your model.

Best of luck with your project, and don't hesitate to reach out if you need further assistance.

Happy training! 🚀
